### PR TITLE
[D1] Add user friendly D1 validation error messages for `dev --remote` and `deploy`

### LINKED ISSUE #4590: [D1] 🚀 Feature Request: Intercept the dev server error "binding DB of type d1 must have a valid `id` specified", make it clearer
### Describe the solution

If you attempt to run `wrangler dev --remote` with `preview_database_id` set to a random string that doesn't exist, you get a big scary error:

```
✘ [ERROR] Error on remote worker: ParseError: A request to the Cloudflare API (/accounts/xxxxxx/workers/scripts/d1-test/edge-preview) failed.

      at throwFetchError
  (/d1-test/node_modules/wrangler/wrangler-dist/cli.js:119636:17)
      at fetchResult
  (/d1-test/node_modules/wrangler/wrangler-dist/cli.js:119555:5)
      at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
      at async createPreviewToken
  (/d1-test/node_modules/wrangler/wrangler-dist/cli.js:148372:29)
      at async createWorkerPreview
  (/d1-test/node_modules/wrangler/wrangler-dist/cli.js:148393:17)
      at async start
  (/d1-test/node_modules/wrangler/wrangler-dist/cli.js:148962:34) {
    text: 'A request to the Cloudflare API
  (/accounts/xxxxxx/workers/scripts/d1-test/edge-preview) failed.',
    notes: [
      {
        text: 'binding DB of type d1 must have a valid `id` specified [code: 10021]'
      }
    ],
    location: undefined,
    kind: 'error',
    code: 10021
  }
```

The only way you'd get this error is if the database doesn't exist, so we should just tell folks to use a real database here.

Attempting to deploy a Worker with a fake ID throws the same error, so we could make that nicer too.

### PR DESCRIPTION
**Pull request by nora-soderlund**
Fixes #4590.

**What this PR solves / how to test:** Adds a user friendly error message mapping for code 10021, specifically when the database id mapping is invalid.

**Author has addressed the following:** I have also moved out some repeated error mappings in the `remote.ts` file to a shared function.

- Tests
  - [ ] Included
  - [X] Not necessary because: Error messages.
- Changeset ([Changeset guidelines](https://github.com/cloudflare/workers-sdk/blob/main/CONTRIBUTING.md#changesets))
  - [X] Included
  - [ ] Not necessary because: 
- Associated docs
  - [ ] Issue(s)/PR(s):
  - [X] Not necessary because: Error messages.

**Note for PR author:**

We want to celebrate and highlight awesome PR review! If you think this PR received a particularly high-caliber review, please assign it the label `highlight pr review` so future reviewers can take inspiration and learn from it.

### CODE REVIEW

**Review by petebacondarwin — COMMENTED** [review]
Thanks for putting this together @nora-soderlund.

I think that it is worth having a changeset for this PR (as a patch update) since it does have user facing consequences, and it is a nice DX improvement that is good to highlight in the changelog.

I think in the long run we should come up with a more general purpose (pluggable) error handling system that doesn't rely so much on switch statements but this seems like a reasonable step forward.

There are some formatting issues picked up in the CI checks... try running:

```
pnpm fix
```

File: packages/wrangler/src/dev/remote.tsx [hunk 1]
```diff
@@ -684,3 +670,58 @@ function ChooseAccount(props: {
...
+ * A switch for handling thrown error mappings to user friendly
+ * messages, does not perform any logic other than logging errors.
+ * @returns if the error was handled or not
+ */
+function handleUserFriendlyError(error: FetchError, accountId?: string) {
+	// simulate aborts as handled errors by this function, but don't
+	// log anything
+	if ((error.code as unknown as string) === "ABORT_ERR") {
+		return true;
+	}
```
[the remark this answers is not in the document]
Comment (nora-soderlund) [reply packages/wrangler/src/dev/remote.tsx]: I'll investigate a bit further, or just revert it to be outside of this function. 👍

File: packages/wrangler/src/deploy/deploy.ts [hunk 2]
```diff
@@ -719,6 +719,10 @@ See https://developers.cloudflare.com/workers/platform/compatibility-dates for m
 				) {
 					err.preventReport();
 
+					if(err.notes[0].text === "binding DB of type d1 must have a valid `id` specified [code: 10021]") {
```
[the remark this answers is not in the document]
Comment (nora-soderlund) [reply packages/wrangler/src/deploy/deploy.ts:722]: It's a good point, I think a more suitable fix is to handle all 10021's as UserError's here.

[the remark this answers is not in the document]
Comment (nora-soderlund) [reply packages/wrangler/src/deploy/deploy.ts:722]: I'm not sure of the best way forwards here actually given the property differences in APIError and UserError; for this PR at least.

I'd say the worst case scenario is acceptable here as is, for now.

File: .changeset/fair-shoes-melt.md [hunk 3]
```diff
@@ -0,0 +1,5 @@
+---
+"wrangler": patch
+---
+
+add user friendly error messages for d1 validation errors
```
Comment (petebacondarwin) [inline .changeset/fair-shoes-melt.md:5]: Following https://github.com/cloudflare/workers-sdk/blob/main/CONTRIBUTING.md#changeset-message-format

Let's add a `fix` type here. Something like:

```suggestion
fix: ensure d1 validation errors render user friendly messages
```

**Review by petebacondarwin — APPROVED** [review]
Assuming CI is happy I think this is a reasonable DX improvement. Thanks!

**Review by rozenmd — APPROVED** [review]
Cheers for picking up that GitHub issue! LGTM.

**Pull request merged**